core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.endpoint;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.client.Endpoint;
+
+/**
+ * A dynamic {@link EndpointGroup}. The list of {@link Endpoint}s can be updated dynamically.
+ */
+public class DynamicEndpointGroup implements EndpointGroup {
+    private final Set<Consumer<Collection<Endpoint>>> updateListeners = new CopyOnWriteArraySet<>();
+    private volatile List<Endpoint> endpoints = ImmutableList.of();
+    private final Lock endpointsLock = new ReentrantLock();
+
+    @Override
+    public final List<Endpoint> endpoints() {
+        return endpoints;
+    }
+
+    /**
+     * Adds the specified {@link Endpoint} to current {@link Endpoint} list.
+     */
+    protected final void addEndpoint(Endpoint e) {
+        try {
+            endpointsLock.lock();
+            ImmutableList.Builder<Endpoint> newEndpointsBuilder = ImmutableList.builder();
+            newEndpointsBuilder.addAll(endpoints);
+            newEndpointsBuilder.add(e);
+            endpoints = newEndpointsBuilder.build();
+            notifyListeners();
```

Comment:
> Would the issue be solved if it's a ReadWriteLock and notify takes a read lock?

No, because we do not want the listeners to be invoked simultaneously?